content/docs/getting-started/hello-features.md

```diff
@@ -0,0 +1,91 @@
+---
+title: "Tokio Features"
+weight : 1020
+menu:
+  docs:
+    parent: getting_started
+---
+
+Feel free to skip this section, and just use `features = ["full"]` while
+learning. This section provides some examples of how and when to specify
+tokio features. 
+
+## Often using all features is fine 
+
+When writing an app, we'll often use all of Tokio's features to accelerate
+our development time:
+
+```toml
+[dependencies]
+tokio = { version = "0.2", features = ["full"] }
+```
+
+For most apps, the additional compile time is hardly noticeable relative to the
+cognitive overhead of thinking about which features are needed.
+
+# Selecting features
+
+When we develop protocol implementations that are used by client apps and
+servers, we want to limit the dependencies to decrease compile time --
+not just for us, but also for the users of our crates.
+
+Let's take a look at our tiny client app and see what features are needed.
+
+## tokio main
+
+If we change our code, so that it only uses the `#[tokio::main]` macro:
+
+```
+#[tokio::main]
+async fn main() {
+    println!("doing nothing yet");
+}
+```
+
+then we can choose whether to use threads (commonly used for servers):
+
+```
+tokio = { version = "0.2", features = ["macros", "rt-threaded"] }
+```
+
+or lightweight cooperative multi-tasking
+(often required for low-profile clients): 
+
+```
+tokio = { version = "0.2", features = ["macros", "rt-core"] }
+```
+
+## TcpStream connect
+
+As we start to build the app, we'll need more features.  The line of code
+where we call `TcpStream::connect` requires two features: 
+
+
+```
+    let stream = TcpStream::connect("127.0.0.1:6142").await.unwrap();
+```
+
+
+1. `dns` for converting the string `127.0.0.1:6142` to an IP address -- Tokio 
```

Comment:
Since this section is talking about using fewer features, should we mention that we can skip `dns` by using an IP address? E.g. here's a version of the code that doesn't need `dns` (and skips the prelude).

src/main.rs:
```rust
use std::net::{IpAddr, Ipv4Addr, SocketAddr};
use tokio::io::AsyncWriteExt;
use tokio::net::TcpStream;

#[tokio::main]
async fn main() {
  let addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 6142);
  let mut stream = TcpStream::connect(addr).await.unwrap();

  let result = stream.write(b"hello world\n").await;
}
```

Cargo.toml:
```toml
[dependencies]
tokio = { version = "0.2", features = ["macros", "rt-threaded", "tcp", "io-util"] }
```

Reply:
this is a great example!  I think `AsyncWriteEx`t vs `prelude` is unrelated, and it could use "rt-core" or "rt-threaded", but the key point is that "dns" can be omitted by specifying `SocketAddr`

```
use std::net::{IpAddr, Ipv4Addr, SocketAddr};
use tokio::prelude::*;
use tokio::net::TcpStream;

#[tokio::main]
async fn main() {
  let addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 6142);
  let mut stream = TcpStream::connect(addr).await.unwrap();

  let _result = stream.write(b"hello world\n").await;
}
```

Reply:
Yeah, the removal of the prelude wasn't necessary but I thought was similarly themed – we are using only what we need instead of the other things the prelude gives you. I also agree with @LucioFranco (I think) in that I think preludes are an unnecessary extra level of indirection. I can see them being useful if it's saving you a lot of boilerplate, but in this case it's just exchanging one line for another.

Reply:
added some narrative to address this.  I referenced some Rust doc I found that says that prelude in crates allows people to easily include whatever is always needed (and I don't think anyone would ever write such a useless app).  In any case, I think it is at least explained.